Adding master election interface to topo.Impl.

alainjobart opened the pull request:
Along with unit test.
Implementing it for zookeeper and etcd.

@michael-berlin @pivanof will work on internal implementation as you guys review this, so if you don't like the interface, please get back to me asap... thanks!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1886)
<!-- Reviewable:end -->

pivanof commented:
Reviewed 9 of 10 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[go/vt/topo/server.go, line 293 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEc0Qwxy0YcR4OW0mm:-KNEc0Qwxy0YcR4OW0mn:1036428733) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L293)):*
> ```Go
> 
> 	//
> 	// Master election, local to multiple cells usually.
> ```

What does "local to multiple cells" mean?

---

*[go/vt/topo/server.go, line 298 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEWKAYhH63XRVniHAH:-KNEWKAZ7BwKnKc-XPHS:-1250378681) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L298)):*
> ```Go
> 	//
> 
> 	// NewParticipation starts participating in the named group,
> ```

s/NewParticipation/NewMasterParticipation/

---

*[go/vt/topo/server.go, line 298 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNF1ImNLZkoxKlKtcas:-KNF1ImObmeNwh5nMszP:1066434441) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L298)):*
> ```Go
> 	//
> 
> 	// NewParticipation starts participating in the named group,
> ```

Both implementations of MasterParticipation don't actually start participating in the election until WaitForMastership() is called, and according to the test that behavior seem to be deliberate. Can you please fix the comment to reflect that?

---

*[go/vt/topo/server.go, line 346 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEX1CWKPJu0fMrbLRS:-KNEX1CXnKJIUTQYublR:-530478446) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L346)):*
> ```Go
> 	// the returned context will be cancelled. If Shutdown was called,
> 	// WaitForMaster will return nil, ErrInterrupted.
> 	WaitForMaster() (context.Context, error)
> ```

WaitForMastership?

---

*[go/vt/topo/test/election.go, line 15 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEaDpY62GTPRrrKGSZ:-KNEaDpY62GTPRrrKGS_:-335075123) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L15)):*
> ```Go
> )
> 
> func waitFormasterID(t *testing.T, mp topo.MasterParticipation, expected string) {
> ```

Please capitalize Master in waitFormasterID.

---

*[go/vt/topo/test/election.go, line 27 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNE_13unJnMtPjf9Z1L:-KNE_13unJnMtPjf9Z1M:799514007) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L27)):*
> ```Go
> 		}
> 
> 		c--
> ```

This seems to be a very primitive and imprecise time measurement. Can we use normal deadlines?

---

*[go/vt/topo/test/election.go, line 32 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNE_QiGNaOrB4sJIhN4:-KNE_QiH1tn3BBj2USVz:-2095649923) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L32)):*
> ```Go
> 		}
> 
> 		t.Logf("Unexpected master %v, expected %v, will retry", master, expected)
> ```

Printing this line 100 times a second is not nice. Can we print it once at the beginning of waiting?

---

*[go/vt/topo/test/election.go, line 37 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNE_vM7pMKesyjZZb_a:-KNE_vM8RjMzfT0jNtmH:-1677921939) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L37)):*
> ```Go
> }
> 
> // checkElection runs the tests on the MasterParticipation part of the API
> ```

The MasterParticipation part of which API? Can you expand the comment?

---

*[go/vt/topo/test/election.go, line 51 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEabWYgiOk51WXA61X:-KNEabWZ0yWqnH7iHiDT:353301005) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L51)):*
> ```Go
> 	waitFormasterID(t, mp1, "")
> 
> 	// wait for it to be the master
> ```

"it" is mp1 here?

---

*[go/vt/topo/test/election.go, line 83 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEcuIa41pglCUa96vq:-KNEcuIbwH5kfrY3batY:-1209950364) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L83)):*
> ```Go
> 
> 	// this should have closed ctx1
> 	timer := time.NewTimer(5 * time.Second)
> ```

Why it is guaranteed that mp1 will close context in 5 seconds?

---

*[go/vt/topo/test/faketopo/faketopo.go, line 20 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEbEus-dGXqhuOv7xK:-KNEbEutaw-2CzwZ2bGU:1117385199) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/faketopo/faketopo.go#L20)):*
> ```Go
> type FakeTopo struct{}
> 
> // Close is part of the topo.Server interface.
> ```

Can these changes be separated at least into a separate commit? They don't have anything to do with MasterParticipation change and there are a lot of these.

---

*[go/vt/zktopo/election.go, line 1 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEmY-9k_GwL7ra0CqG:-KNEmY-At_NNUofIh5z1:-517706281) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/zktopo/election.go#L1)):*
> ```Go
> // Copyright 2013, Google Inc. All rights reserved.
> ```

2016?

---

*[go/vt/zktopo/election.go, line 114 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEorytf_7QBRjakzRq:-KNEorytf_7QBRjakzRr:1953083601) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/zktopo/election.go#L114)):*
> ```Go
> //   being the master.
> // - wait for mp.shutdown.
> func (mp *zkMasterParticipation) whileMaster(proposal string, cancel context.CancelFunc) {
> ```

"whileMaster" seem to be a bad function name not indicating the purpose of the function. Not sure what's the best name would be, but maybe WaitForShutdown() or HoldMastership()?

---

*[go/zk/zkconn.go, line 18 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEp6bag69isGkVvrAF:-KNEp6bag69isGkVvrAG:23781928) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/zk/zkconn.go#L18)):*
> ```Go
> 	log "github.com/golang/glog"
> 	zookeeper "github.com/samuel/go-zookeeper/zk"
> 	"golang.org/x/net/context"
> ```

Why this was moved?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
Review status: 6 of 10 files reviewed at latest revision, 14 unresolved discussions.

---

*[go/vt/topo/server.go, line 293 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEc0Qwxy0YcR4OW0mm:-KNIXlZDLVRSVEPNL0bt:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L293)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> What does "local to multiple cells" mean?
</details>
Done.

---

*[go/vt/topo/server.go, line 298 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEWKAYhH63XRVniHAH:-KNIYFYr_dEQOAI7IWjs:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L298)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> s/NewParticipation/NewMasterParticipation/
</details>
Done.

---

*[go/vt/topo/server.go, line 298 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNF1ImNLZkoxKlKtcas:-KNIYGHVPhFu23DKJrGx:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L298)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Both implementations of MasterParticipation don't actually start participating in the election until WaitForMastership() is called, and according to the test that behavior seem to be deliberate. Can you please fix the comment to reflect that?
</details>
Done.

---

*[go/vt/topo/server.go, line 346 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEX1CWKPJu0fMrbLRS:-KNIYQwXcV7ldh27o-Ot:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L346)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> WaitForMastership?
</details>
Done.

---

*[go/vt/topo/test/election.go, line 15 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEaDpY62GTPRrrKGSZ:-KNIYs51xNNCGA0GvWor:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L15)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Please capitalize Master in waitFormasterID.
</details>
Done.

---

*[go/vt/topo/test/election.go, line 27 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNE_13unJnMtPjf9Z1L:-KNIZiVTFHu7DD8Z9jFh:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L27)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> This seems to be a very primitive and imprecise time measurement. Can we use normal deadlines?
</details>
Done.

---

*[go/vt/topo/test/election.go, line 32 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNE_QiGNaOrB4sJIhN4:-KNIZrRp5278nkjObno5:830726044) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L32)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Printing this line 100 times a second is not nice. Can we print it once at the beginning of waiting?
</details>
I just needed printed it for debugging when I was coding it, removing it.

---

*[go/vt/topo/test/election.go, line 37 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNE_vM7pMKesyjZZb_a:-KNI_0wkeHU8GXoT_b0A:-1950587741) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L37)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> The MasterParticipation part of which API? Can you expand the comment?
</details>
This whole package is testing topo.Impl API, so that's kinda implicit. Extended the comment anyway.

---

*[go/vt/topo/test/election.go, line 51 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEabWYgiOk51WXA61X:-KNI_EyulX0_YickuXqe:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L51)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> "it" is mp1 here?
</details>
Done.

---

*[go/vt/topo/test/election.go, line 83 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEcuIa41pglCUa96vq:-KNI_Xf_H7-3g7NVNqfo:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L83)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Why it is guaranteed that mp1 will close context in 5 seconds?
</details>
Done.

---

*[go/vt/topo/test/faketopo/faketopo.go, line 20 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEbEus-dGXqhuOv7xK:-KNI_nlr0Df8HrlQhjzq:-32198068) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/faketopo/faketopo.go#L20)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Can these changes be separated at least into a separate commit? They don't have anything to do with MasterParticipation change and there are a lot of these.
</details>
A bit annoying to separate them now. I discovered that when I was adding the new API, so the file was already opened in my change...

---

*[go/vt/zktopo/election.go, line 1 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEmY-9k_GwL7ra0CqG:-KNI_mg2LJ_e-yde1TVF:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/zktopo/election.go#L1)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> 2016?
</details>
Done.

---

*[go/vt/zktopo/election.go, line 114 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEorytf_7QBRjakzRq:-KNIaMNzYrP4K3R_VRBF:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/zktopo/election.go#L114)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> "whileMaster" seem to be a bad function name not indicating the purpose of the function. Not sure what's the best name would be, but maybe WaitForShutdown() or HoldMastership()?
</details>
Done.

---

*[go/zk/zkconn.go, line 18 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEp6bag69isGkVvrAF:-KNIaRbBdwyecVPEIYgJ:1637859170) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/zk/zkconn.go#L18)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Why this was moved?
</details>
By convention,  we should have all standard go imports in a section, then all non-vitess ones, then all internal ones. Just fixing that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886)*
<!-- Sent from Reviewable.io -->

pivanof commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[go/vt/topo/server.go, line 298 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNF1ImNLZkoxKlKtcas:-KNIev58WBm5TqMW95zD:1464084597) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L298)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> Done.
</details>
Can you still add a comment about when the participation in the election is supposed to start? Either here or at WaitForMastership() comment.

---

*[go/vt/topo/server.go, line 295 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIdpbexSaX7slz5ie7:-KNIdpbexSaX7slz5ie8:1757299095) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L295)):*
> ```Go
> 	// Master election methods. This is meant to have a small number of
> 	// processes elect a master within a group. The backend storage for this
> 	// can either be the global topo server, or a chorum of local cells.
> ```

I can't find translation of the word "chorum". What does it mean?

---

*[go/vt/topo/server.go, line 349 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIhfjxIk7_R7ocKHEP:-KNIhfjxIk7_R7ocKHEQ:966699810) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L349)):*
> ```Go
> 	WaitForMastership() (context.Context, error)
> 
> 	// Shutdown is called when we don't want to participate in the
> ```

I've just realized that this API assumes that participation in the master election is always started at the start of the application and ends at shutdown. But it's not always true, right? Even the comment here says "called when we don't want to participate", that doesn't necessarily means it's a shutdown of the application. You can argue that it's a shutdown of participation, but I think that's not really an appropriate word for that. Besides that actually it's not clear now when participation in the master election starts and when it ends. It looks like your intention was that participation starts on the first call to WaitForMastership(), but then when the returned context is canceled (i.e. we lost mastership) does participation continues? Or should it stop? Can you clarify with comments where the intended boundaries are and clarify the API itself for that, e.g. Shutdown() should probably be Stop().

---

*[go/vt/topo/test/election.go, line 83 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEcuIa41pglCUa96vq:-KNIkRETaHO0nBzMIPm9:-59535576) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L83)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> Done.
</details>
So it's part of API guarantee that it must always close the context within 5 seconds of the call to Shutdown()? If yes can you add this requirement to the API comments, so that anyone implementing it knew what they have to provide?

---

*[go/vt/topo/test/faketopo/faketopo.go, line 20 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEbEus-dGXqhuOv7xK:-KNIl-BL4mmUm_cVOUuX:1694449755) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/faketopo/faketopo.go#L20)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> A bit annoying to separate them now. I discovered that when I was adding the new API, so the file was already opened in my change...
</details>
Why annoying? Just create a commit with these changes and rebase this branch on top of that commit.

---

*[go/zk/zkconn.go, line 18 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEp6bag69isGkVvrAF:-KNIlSb0aV5E4oF2-Hfa:-1141761512) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/zk/zkconn.go#L18)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> By convention,  we should have all standard go imports in a section, then all non-vitess ones, then all internal ones. Just fixing that.
</details>
Can you extract this into a different PR or at least different commit? It's completely unrelated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
Review status: 6 of 10 files reviewed at latest revision, 16 unresolved discussions.

---

*[go/vt/topo/server.go, line 298 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNF1ImNLZkoxKlKtcas:-KNIotVb7usQKg3ITYLG:-499146009) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L298)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Can you still add a comment about when the participation in the election is supposed to start? Either here or at WaitForMastership() comment.
</details>
Done.

---

*[go/vt/topo/server.go, line 295 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIdpbexSaX7slz5ie7:-KNIoW1j1jpjR-r8B_Oc:782744752) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L295)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> I can't find translation of the word "chorum". What does it mean?
</details>
quorum. dunno why I spelled it so wrong...

---

*[go/vt/topo/server.go, line 349 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIhfjxIk7_R7ocKHEP:-KNIpX18g1auE2TgnN4T:1939983622) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L349)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> I've just realized that this API assumes that participation in the master election is always started at the start of the application and ends at shutdown. But it's not always true, right? Even the comment here says "called when we don't want to participate", that doesn't necessarily means it's a shutdown of the application. You can argue that it's a shutdown of participation, but I think that's not really an appropriate word for that. Besides that actually it's not clear now when participation in the master election starts and when it ends. It looks like your intention was that participation starts on the first call to WaitForMastership(), but then when the returned context is canceled (i.e. we lost mastership) does participation continues? Or should it stop? Can you clarify with comments where the intended boundaries are and clarify the API itself for that, e.g. Shutdown() should probably be Stop().
</details>
Added more comments and renamed to Stop, you're right it's clearer.

---

*[go/vt/topo/test/election.go, line 83 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEcuIa41pglCUa96vq:-KNIqC39Mfq63axFtRId:-649670664) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/test/election.go#L83)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> So it's part of API guarantee that it must always close the context within 5 seconds of the call to Shutdown()? If yes can you add this requirement to the API comments, so that anyone implementing it knew what they have to provide?
</details>
the comment should say 'as soon as possible', there is no time expectation. 5s is just the test timing out when it becomes unreasonable.

---

*[go/zk/zkconn.go, line 18 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEp6bag69isGkVvrAF:-KNIqVmtDWiVKGGWHsz5:-1659755444) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/zk/zkconn.go#L18)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> Can you extract this into a different PR or at least different commit? It's completely unrelated.
</details>
so where do we stop and create different commits here? a typo in a comment ten lines earlier should also be a different commit? this is a bit extreme, I'd rather fix very minor stuff like this as I find it and just take it as part of the commit... then event though extracting it to a different commit is possible, it feels like a huge waste of time... so when it's convenient, I'll do it, but this is a bit late for this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Sorry that it took me so long to review it.

I'm fine with the API and the different implementations after my comments are addressed.

Overall, I think this is the right way to do it and we can refine it if necessary along the way.

---

Reviewed 6 of 10 files at r1, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 28 unresolved discussions.

---

*[go/vt/etcdtopo/election.go, line 32 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKV7BjFFr-nCUqJarD:-KNKV7BjFFr-nCUqJarE:602447395) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/etcdtopo/election.go#L32)):*
> ```Go
> // etcdMasterParticipation implements topo.MasterParticipation.
> //
> // We create a single file and use etcd's compare&swap with TTLs.
> ```

Please mention here that the lock file will be per `name` and that the content of the file will be the `id` of the current master.

---

*[go/vt/etcdtopo/election.go, line 107 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKVp4XjgAdittoCdaE:-KNKVp4XjgAdittoCdaF:-65186862) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/etcdtopo/election.go#L107)):*
> ```Go
> 		// Wait for the lock file to be deleted, then try again.
> 		ctx, cancel := context.WithCancel(context.Background())
> 		go func() {
> ```

Add a comment here that the sole purpose of this go routine is to cancel out the wait if Stop() is called.

---

*[go/vt/topo/server.go, line 346 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEX1CWKPJu0fMrbLRS:-KNKuYhwUlRXi4oHdzbH:-756568132) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L346)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> Done.
</details>
How about `StartAndWaitForMastership`? This would be in line with the new `Stop` method.

Alternatively, something shorter e.g. `StartAndWait`.

---

*[go/vt/topo/server.go, line 295 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIdpbexSaX7slz5ie7:-KNIrCL5Bj9V7JihPcka:1534824774) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L295)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> quorum. dunno why I spelled it so wrong...
</details>
Can you please explain what you mean with `local cells`? I interpret `cell` with `data center` and I don't think that makes sense here?

---

*[go/vt/topo/server.go, line 302 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIrzEW0umUaHDAAiH8:-KNIrzEW0umUaHDAAiH9:-498367101) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L302)):*
> ```Go
> 	// Id is the name of the local process, passing in the hostname:port of the
> 	// current process as id is the common usage.
> 	NewMasterParticipation(name, id string) (MasterParticipation, error)
> ```

Please clarify in the comments that `id` must be unique between different (running) master processes for the same `name`.

---

*[go/vt/topo/server.go, line 345 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIsLJLa4E7UubUW3pO:-KNIsLJLa4E7UubUW3pP:-1595811709) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L345)):*
> ```Go
> 	// WaitForMastership waits until this process is the master.
> 	// After we become the master, we may lose mastership. In that case,
> 	// the returned context will be cancelled. If Shutdown was called,
> ```

Please write `canceled` with a single `l`. The Go team recently switched to this spelling. Therefore, let's try to use that spelling as well.

---

*[go/vt/topo/server.go, line 349 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIhfjxIk7_R7ocKHEP:-KNItaFGGeizsslS_XHV:2058909501) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L349)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> Added more comments and renamed to Stop, you're right it's clearer.
</details>
I was about to suggest `Release` or `Relinquish` :)

But I think we should use `Stop` because the internal API uses that as well.

---

*[go/vt/topo/test/election.go, line 87 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIwSVPXEXkGyHrHdpK:-KNIwSVQdyeP_Mh6nTzX:1547928149) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/topo/test/election.go#L87)):*
> ```Go
> 	select {
> 	case <-ctx1.Done():
> 	case <-timer.C:
> ```

You can use `time.After(5 * time.Second)` here instead and get rid of the `timer` object.

---

*[go/vt/topo/test/testing.go, line 96 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIz841mYKIg2cF2GWF:-KNIz842afiyRiZsbMV-:1935097971) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/topo/test/testing.go#L96)):*
> ```Go
> 	ts.Close()
> 
> 	t.Log("=== checkElection")
> ```

I'm not a big fan of such banner comments but I understand the need for it here.

Maybe add another one on top of `checkVSchema` as well to be consistent?

---

*[go/vt/zktopo/election.go, line 51 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKFxGyvjikp-c0zMEd:-KNKFxGzvnqowlnL3BYI:1792875415) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L51)):*
> ```Go
> // zkMasterParticipation implements topo.MasterParticipation.
> //
> // We use a directory with files created as sequence and ephemeral,
> ```

Please also mention here that there will be one directory per `name`.

---

*[go/vt/zktopo/election.go, line 77 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKH_l9xwQViV3Dl5VT:-KNKH_l9xwQViV3Dl5VU:-327926968) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L77)):*
> ```Go
> 	select {
> 	case <-mp.stop:
> 		close(mp.done)
> ```

Multiple calls to `WaitForMastership` after `Stop` will panic here because the channel is already closed.

If we say that you should not call `WaitForMastership` after `Stop`, that's fine with me. In that case, you may need to clarify the API description in `server.go`.

There are multiple call sites `close(mp.done)` in this function and `watchMastership` which make it hard to reason whether we actually caught all cases.

What about the following alternative: Use a watch group which you increment in this function and also before calling `go mp.watchMastership`? You can add a defer to both functions as well. The count changes would look like this:

```
WaitForMastership() -> 1
go mp.watchMastership() -> 2
defer from WaitForMasterShip() -> 1
defer from watchMastership() -> 0
```

(The same comment applies to the etcd implementation as well.)

---

*[go/vt/zktopo/election.go, line 90 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKLSqLJOjK_Snlvg9B:-KNKLSqLJOjK_Snlvg9C:1987176743) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L90)):*
> 24*time.Hour

Let's use a shorter timeout here to make sure that the timeout retry code gets exercised during manual testing or as part of canarying? e.g. something like 5 minutes?

---

*[go/vt/zktopo/election.go, line 127 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKJKnMZ7AoXCCSRGhq:-KNKJKnMZ7AoXCCSRGhr:-915026462) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L127)):*
> ```Go
> 	_, stats, events, err := mp.zkts.zconn.GetW(proposal)
> 	if err != nil {
> 		log.Warningf("Cannot watch proposal whle being master, stopping: %v", err)
> ```

typo: while

---

*[go/vt/zktopo/election.go, line 139 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKJZ4iR3xzDqLHqm_A:-KNKJZ4jhiPg0BcTzm7h:1622838898) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L139)):*
> ```Go
> 			log.Warningf("Error deleting our proposal %v: %v", proposal, err)
> 		}
> 		close(mp.done)
> ```

How about calling this by a `defer`? e.g. I'm wondering if the case below should close `mp.done` as well.

---

*[go/vt/zktopo/election.go, line 153 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKN4PTrv9K7TGIz2VG:-KNKN4PTrv9K7TGIz2VH:1581498691) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L153)):*
> ```Go
> }
> 
> // GetCurrentMasterID is part of the topo.MasterParticipation interface
> ```

I would prefer a short comment here where you clarify that the node with the smallest sequence number in its name is the current master.

---

*[go/zk/zkconn.go, line 18 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEp6bag69isGkVvrAF:-KNKDevLs1mbUYESUHKB:-738992641) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/zk/zkconn.go#L18)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> so where do we stop and create different commits here? a typo in a comment ten lines earlier should also be a different commit? this is a bit extreme, I'd rather fix very minor stuff like this as I find it and just take it as part of the commit... then event though extracting it to a different commit is possible, it feels like a huge waste of time... so when it's convenient, I'll do it, but this is a bit late for this one.
</details>
I agree with Alain on this one. We should not enforce this for our pull requests. It should be up to the author, not the reviewer to factor out small changes like this.

In the past I did this very often (Eclipse has a nice Git Staging view which makes it very easy to move changed parts of a file to the Git index). But after all, I found that it's not worth the time. Therefore, I'll probably do this less and less.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886#-:-KNKtceejuvmYZsxhCre:-501966947)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
Review status: all files reviewed at latest revision, 28 unresolved discussions.

---

*[go/vt/etcdtopo/election.go, line 32 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKV7BjFFr-nCUqJarD:-KNXJVlRDuJJtD3WLN-3:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/etcdtopo/election.go#L32)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please mention here that the lock file will be per `name` and that the content of the file will be the `id` of the current master.
</details>
Done.

---

*[go/vt/etcdtopo/election.go, line 107 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKVp4XjgAdittoCdaE:-KNXK7B0MQXbWbnQdAvF:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/etcdtopo/election.go#L107)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Add a comment here that the sole purpose of this go routine is to cancel out the wait if Stop() is called.
</details>
Done.

---

*[go/vt/topo/server.go, line 346 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEX1CWKPJu0fMrbLRS:-KNXKtJEjUWBq3aiqol1:-1097870222) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L346)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> How about `StartAndWaitForMastership`? This would be in line with the new `Stop` method.
> 
> Alternatively, something shorter e.g. `StartAndWait`.
</details>
It's not symmetrical to Stop, actually. WaitForMastership is called in a loop, Stop is called at the end. So I don't think we should have Start in the name... NewMasterParticipation is the Stop() equivalent.

---

*[go/vt/topo/server.go, line 295 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIdpbexSaX7slz5ie7:-KNXKTuSaMway5M8wc29:-499146009) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L295)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Can you please explain what you mean with `local cells`? I interpret `cell` with `data center` and I don't think that makes sense here?
</details>
Done.

---

*[go/vt/topo/server.go, line 302 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIrzEW0umUaHDAAiH8:-KNXKeXHcY01SdHWGuGu:-499146009) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L302)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please clarify in the comments that `id` must be unique between different (running) master processes for the same `name`.
</details>
Done.

---

*[go/vt/topo/server.go, line 345 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIsLJLa4E7UubUW3pO:-KNXKiYAAhFSIpTjbh5W:-97981778) ([raw file](https://github.com/youtube/vitess/blob/d4b15ffa316ed0aed27f1599d54e84a11c07e827/go/vt/topo/server.go#L345)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please write `canceled` with a single `l`. The Go team recently switched to this spelling. Therefore, let's try to use that spelling as well.
</details>
I'm trying, I think it's the only place I missed it. ;)

---

*[go/vt/topo/test/election.go, line 87 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIwSVPXEXkGyHrHdpK:-KNXLhNGkv8aejvhnWJ-:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/topo/test/election.go#L87)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> You can use `time.After(5 * time.Second)` here instead and get rid of the `timer` object.
</details>
Done.

---

*[go/vt/topo/test/testing.go, line 96 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIz841mYKIg2cF2GWF:-KNXLtLpF12yFHwLJhTF:1939806945) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/topo/test/testing.go#L96)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> I'm not a big fan of such banner comments but I understand the need for it here.
> 
> Maybe add another one on top of `checkVSchema` as well to be consistent?
</details>
it is there?

---

*[go/vt/zktopo/election.go, line 51 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKFxGyvjikp-c0zMEd:-KNXMCpn2jsGmbDKDWKe:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L51)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please also mention here that there will be one directory per `name`.
</details>
Done.

---

*[go/vt/zktopo/election.go, line 77 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKH_l9xwQViV3Dl5VT:-KNXMIcc3PUSLqQpm-vq:-925120384) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L77)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Multiple calls to `WaitForMastership` after `Stop` will panic here because the channel is already closed.
> 
> If we say that you should not call `WaitForMastership` after `Stop`, that's fine with me. In that case, you may need to clarify the API description in `server.go`.
> 
> There are multiple call sites `close(mp.done)` in this function and `watchMastership` which make it hard to reason whether we actually caught all cases.
> 
> What about the following alternative: Use a watch group which you increment in this function and also before calling `go mp.watchMastership`? You can add a defer to both functions as well. The count changes would look like this:
> 
> ```
> WaitForMastership() -> 1
> go mp.watchMastership() -> 2
> defer from WaitForMasterShip() -> 1
> defer from watchMastership() -> 0
> ```
> 
> (The same comment applies to the etcd implementation as well.)
</details>
Using a WaitGroup is a great alternative, I will make that change. In a separate PR though, will be easier for me. Already have had this PR in flight for a long time. :)

---

*[go/vt/zktopo/election.go, line 90 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKLSqLJOjK_Snlvg9B:-KNXN4x8ElbA_ISrhiQf:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L90)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> > 24*time.Hour
> 
> Let's use a shorter timeout here to make sure that the timeout retry code gets exercised during manual testing or as part of canarying? e.g. something like 5 minutes?
</details>
Done.

---

*[go/vt/zktopo/election.go, line 127 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKJKnMZ7AoXCCSRGhq:-KNXNA6kErBvcyX9mhVk:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L127)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> typo: while
</details>
Done.

---

*[go/vt/zktopo/election.go, line 139 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKJZ4iR3xzDqLHqm_A:-KNXNKZwwjRiJzlaG55_:-1323166348) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L139)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> How about calling this by a `defer`? e.g. I'm wondering if the case below should close `mp.done` as well.
</details>
it should not, because below it is called when we lose mastership, not when stop() was called.

---

*[go/vt/zktopo/election.go, line 153 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNKN4PTrv9K7TGIz2VG:-KNXNfp-r6XA-__75ZTV:-499146009) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/zktopo/election.go#L153)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> I would prefer a short comment here where you clarify that the node with the smallest sequence number in its name is the current master.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[go/vt/etcdtopo/election.go, line 112 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNXwJOtEqH-TF9hcasl:-KNXwJOtEqH-TF9hcasm:1520529433) ([raw file](https://github.com/youtube/vitess/blob/a03efa927e6ba22c94fbb78c3240b83a45a0db83/go/vt/etcdtopo/election.go#L112)):*
> ```Go
> 		go func() {
> 			// This go routine cancels out the context if Stop()
> 			// it called.
> ```

typo: is

---

*[go/vt/topo/server.go, line 346 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNEX1CWKPJu0fMrbLRS:-KNXx9iDBziQ9tVjc_JN:-1501033631) ([raw file](https://github.com/youtube/vitess/blob/ecdf8430c22b82675e1d9696f1a78b84556110a6/go/vt/topo/server.go#L346)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> It's not symmetrical to Stop, actually. WaitForMastership is called in a loop, Stop is called at the end. So I don't think we should have Start in the name... NewMasterParticipation is the Stop() equivalent.
</details>
`NewMasterParticipation` only initializes the object but e.g. doesn't interact with the topology yet.

This is delayed until this method. IMO, this method does two things: It starts getting the master lock and it also blocks until the lock is actually obtained. If you would split this method into two methods, they would be called `Start` and `Wait`. Hence the proposal `StartAndWait`.

Also: After `NewMasterParticipation` is called there is no need to call `Stop`. But after this method is called (`StartAndWait...` :)), `Stop` must be called.

---

*[go/vt/topo/server.go, line 302 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNXwba1ZPyUhBCJKsd3:-KNXwba1ZPyUhBCJKsd4:-1752326639) ([raw file](https://github.com/youtube/vitess/blob/a03efa927e6ba22c94fbb78c3240b83a45a0db83/go/vt/topo/server.go#L302)):*
> ```Go
> 	// NewMasterParticipation creates a MasterParticipation
> 	// object, used to become the Master in an election for the
> 	// provided group name.  Id is the name of the local process,
> ```

nit: double space

---

*[go/vt/topo/test/testing.go, line 96 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1886#-KNIz841mYKIg2cF2GWF:-KNXz8XfMx7d5_DNTUuE:-1067966250) ([raw file](https://github.com/youtube/vitess/blob/50c41614e49df24df755f826cdf2d24295d99e0a/go/vt/topo/test/testing.go#L96)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> it is there?
</details>
Sorry, the diff view fooled me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1886)*
<!-- Sent from Reviewable.io -->
